### DISCUSSION TITLE
[PROPOSAL] Error message design guidelines

### DISCUSSION THREAD

**Discussion by ghost** [apache/superset#19769] [Ideas]
Hi all, 

This is a proposal to add [design guidelines](https://github.com/apache/superset/wiki/Design-Guidelines) to the Wiki detailing best practices around crafting error messages. Note that no new UI is being proposed, just guidelines around using patterns established during [SIP-40](https://github.com/apache/superset/issues/9194).

----

# Error message design guidelines

### Definition
Interface errors appear when the application can’t do what the user wants, typically because:
	•	The app technically fails to complete the request
	•	The app can’t understand the user input
	•	The user tries to combine operations that can’t work together

In all cases, encountering errors increases user friction and frustration while trying to use the application. Providing an error experience that helps the user understand what happened and their next steps is key to building user confidence and increasing engagement.

### General best practices
**The best error experience is no error at all.** Before implementing error patterns, consider what you might do in the interface before the user would encounter an error to prevent it from happening at all. This might look like:
	•	Providing tooltips or microcopy to help users understand how to interact with the interface
	•	Disabling parts of the UI until desired conditions are met (e.g. disabling a Save button until mandatory fields in a form are completed)
	•	Correcting an error automatically (e.g. autocorrecting spelling errors)

**Only report errors users care about.** The only errors that should appear in the interface are errors that require user acknowledgement and action (even if that action is “try again later” or “contact support”).

**Do not start the user in an error state.** If user inputs are required to display an initial interface (e.g. a chart in Explore), use empty states or field highlighting to drive users to the required action.

### Pattern selection

Select one pattern per error (e.g. do not implement an inline and banner pattern for the same error).

| When the error... | Use... |
| :--- | :--- 
| Is directly related to a UI control (user didn’t fill out a mandatory field, user inputs wrong datatype, etc.) | Inline error |
| Is not directly related to a UI control (technical failure, loading issues, etc.) | Banner error |

### Inline
Inline errors are used when the source of the error is directly related to a UI control (text input, selector, etc.) such as the user not populating a required field or entering a number in a text field.

**Anatomy**
Use the `LabeledErrorBoundInput` component for this error pattern.

<img width="386" alt="Inline" src="https://user-images.githubusercontent.com/90715511/163874643-eb3c65f4-6b5f-42de-b370-73a0c6a8aa6e.png">

1. **Message**
Provides direction on how to populate the input correctly.

**Implementation details**
	•	Where and when relevant, scroll the screen to the UI containing the inline error
	◦	If form has multiple fields with inline errors, scroll user to the error closest to the top of the screen and let the user scroll to the next error

### Banner
Banner errors are used when the source of the error is not directly related to a UI control, but reflects a larger error related to an area of the screen or the entire screen or application.

Place the component as close to the relevant area as possible (e.g. if the error pertains to a chart loading in Explore, display banner where chart is expected to load). If the error happens in a modal (and is not directly tied to a field), display banner at the bottom, in the footer (see example). If the error is app-wide, place the banner at the top of the screen (above the navigation).

**Anatomy**
Use the `ErrorAlert` component for this error pattern.

Error banner collapsed
<img width="1455" alt="Banner-Collapsed" src="https://user-images.githubusercontent.com/90715511/163875122-91871d0c-10b5-4ced-883e-6cc147eb0a24.png">

Error banner expanded
<img width="1455" alt="Banner-Expanded" src="https://user-images.githubusercontent.com/90715511/163875180-54c24754-c9ba-46aa-b969-ce4e8dce7c7c.png">

Modal opened from banner error
<img width="1455" alt="Banner-Modal" src="https://user-images.githubusercontent.com/90715511/163875198-c8348bd5-e124-4275-9614-75215229cd25.png">

Alert component in footer
<img width="733" alt="Screen Shot 2022-04-18 at 6 12 21 PM" src="https://user-images.githubusercontent.com/90715511/163900604-ec1995c8-d820-430b-b3ec-796d9b2dc347.png">

1. **Headline [optional]**
One or two words that summarize the type of error. Can omit if there are space constraints.

2. **Message**
Explains what went wrong and user next steps.

3. **Message expand [optional]**
Expands banner to reveal message details when necessary. Note: This text should always say “See more” (to expand) and “See less” (to collapse).

4. **Message details [optional]**
Provides additional, useful information that may help the user resolve the error. This additional text should follow content guidelines and should not be required for the user to resolve the error.

5. **Modal [optional]**
It may not always be possible to expand an error message in a banner for spatial reasons (e.g. an individual chart banner error in a dashboard). In these cases, a “See more” button may open a modal containing the error message instead. This pattern should be used only if absolutely necessary. Use the `ToastType.DANGER` component for this error pattern.

**Implementation details**
	•	Banner remains on the screen until the error is resolved
	•	Banner should move other screen content, not overlay it

## Content
Useful error messages communicate two things to the user:
1. What went wrong
2. What the user should do next

### Guidelines
Useful error messages:
	•	Are clear and accurate, leaving no room for misinterpretation
	•	Are short and meaningful
	•	Can be understood by non-technical users
	•	Don’t blame the user
	•	Avoid using negative words

Don’t forget to follow [capitalization guidelines](https://github.com/apache/superset/wiki/Capitalization-Guidelines).

**Examples**

| 🛑 Don’t  | ✅ Do |
| :--- | :--- 
| “Cannot delete a datasource that has slices attached to it.” | “Please delete all charts using this dataset before deleting the dataset. [list charts].” |
| “You cannot give two datasets the same name.” | “A dataset already exists with this name. Please give a unique name to this dataset.” |
| “System failure.” | “Superset encountered an error while trying to load your dashboard. Please try again later.” |

**Comment by cccs-rc** [comment]
Would these guidelines also be a good place to mention the importance of making sure the more technical details related to an error are written to appropriate logs so that developers/system admins can easily debug? We've had a few instances where a concise error message in the UI, such as "Superset encountered an error", didn't spawn any further details in the logs (presumably because the error message was getting swallowed somewhere), which left us scratching our heads for a while!

**Reply by ghost** [reply]
These are specifically design guidelines, so I'm not sure it would be the best place for that – @rusackas where would the technical guidelines live?

**Reply by rusackas** [reply]
The closest thing I can think of is [SIP-40](https://github.com/apache/superset/issues/9194) which goes into how this ought to work, but while a plan was voted in, there's more to do in terms of both implementation and documentation. If anyone has bandwidth to write this up as a proper Code Style Guideline and/or drive the SIP further toward completion, I think we'd all be thankful for that :D
